docs/rfcs/004-multi-file-diff.md

```diff
@@ -0,0 +1,67 @@
+
+#### Staged Changes pane
+- Same behavior as Unstaged Changes pane.
+- Once there is at least one file selected, `Unstage All` button should be worded as `Unstage Selected`.
+
+#### Multi-file diff view
+_(note: The following is a summary of what we would like the UX to achieve, but I don't have a clear visuals of what that looks like yet.)_
+
+- Shows diffs of multiple files as a stack.
+- Each diff should show up as its own block, and the current functionality should remain independent of each block.
+- It should be easy to jump quickly to a specific file you care about, or back to the file list to get to another file. Dotcom does so by creating a `jump to` drop down.
+- As user scrolls through a long list of diffs, there should be a sticky heading which remains visible showing the filename of the diff being viewed.
+Nice-to-have UX that doesn't necessarily need to be implemented
+- Each file diff can be collapsed (this can potentially be )
+
+
+## Drawbacks
+
+- `cmd-click` to select multiple files might not be as universally known as we assume, so that might affect discoverability of this feature.
+- There might be performance concerns having to render many diffs at once.
+
+## Rationale and alternatives
+
+An alternative would be to _not_ implement multi-file diff, as other editors like VS Code also only has per-file diff at the time of writing. However, not implementing this would imply that [the proposed new PR review flow](https://github.com/atom/github/blob/master/docs/rfcs/003-pull-request-review.md) will have to find another solution to display all changes in a PR. Additionally users would have to do a lot more clicking to view all of their changes. Imagine there was a variable rename and only 10 lines are changed, but they are each in a different file. It'd be a bit of a pain to click through to view each one. Also, if we didn't implement multi-file diffs then we couldn't show commit contents since they often include changes across multiple files.
+
+## Unresolved questions
+
+- What unresolved questions do you expect to resolve through the implementation of this feature before it is released in a new version of the package?
+
+How exactly do we construct the multi-file diffs? Do we have one TextEditor component that has different sections for each file. Or do we create a new type of pane item that contains multiple TextEditor components stacked on top of one another, one for each file diff... If we do the former we could probably get something shipped sooner (we could just get the diff of the staged changes from Git, add a special decoration for file headers, and present all the changes in one editor). But to pave the way for a more complex code review UX I think taking extra time to do the latter will serve us well. For example, I can imagine reviewers wanting to collapse some files, or mark them as "Done", in which case it would be easier if we treated each diff as its own component.
```

Comment:
I agree - it seems more robust to have each file be its own `TextEditor`. It also seems like editable diffs would be easier if we go this route.  If we have one large `TextEditor` that has data from multiple files, when you save a buffer, we have to map that to which file that data lives in, which seems annoying.

Reply:
> [..] If we do the former we could probably get something shipped sooner [..]

Actually, I think it'd be _easier_ to implement the multiple-`TextEditor` approach:

```js
<div className="github-MultiFilePatchView">
  <AtomTextEditor /* file one */ />
  <AtomTextEditor /* file two */ />
  <AtomTextEditor /* file three */ />
</div>
```

Using one `<AtomTextEditor>` to render all patches, with the file headers as block decorations, would involve some more complicated fiddling in the model layer to keep everything straight - even more-so for editability, as @annthurium noted above.

The tradeoff I see is that the single `TextEditor` approach is likely to scale better to patches with large numbers of files - by pushing complexity into the decorations within the `TextEditor`, we gain the performance boost associated with the editor's tiling and scrolling optimizations.